src/Chapter3.hs

```diff
@@ -1006,6 +1142,189 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- Couldn't make it simpler :(
+data KnightAction
+  = KnightAttack
+  | Drink
+  | Cast
+  deriving (Show)
+
+data MonsterAction
+  = MonsterAttack
+  | Run
+  deriving (Show)
+
+data Knight = Knight
+  { knightHealth :: Int
+  , knightAttack :: Int
+  , knightDefense :: Int
+  , knightActions :: [KnightAction]
+  , baseKnightHealth :: Int
+  , baseKnightHealthIncrease :: Int
+  , baseKnightDefense :: Int
+  , baseKnightDefenseIncrease :: Int
+  } deriving (Show)
+
+data Monster = Monster
+  { monsterHealth :: Int
+  , monsterAttack :: Int
+  , monsterActions :: [MonsterAction]
+  } deriving (Show)
+
+data ActionType
+  = GenericAttack
+  | SpecialAction
+  deriving (Show)
+
+getFirstKnightAction :: Knight -> KnightAction
+getFirstKnightAction Knight { knightActions = (x:_) } = x
+getFirstKnightAction _ = error "Forgot to 'cycle' knight actions"
+
+getFirstMonsterAction :: Monster -> MonsterAction
+getFirstMonsterAction Monster { monsterActions = (x:_) } = x
+getFirstMonsterAction _ = error "Forgot to 'cycle' monster actions"
+
+getFollowingKnightActions :: Knight -> [KnightAction]
+getFollowingKnightActions Knight { knightActions = (_:xs) } = xs
+getFollowingKnightActions _ = error "Forgot to 'cycle' knight actions"
+
+getFollowingMonsterAction :: Monster -> [MonsterAction]
+getFollowingMonsterAction Monster { monsterActions = (_:xs) } = xs
+getFollowingMonsterAction _ = error "Forgot to 'cycle' monster actions"
```

Comment:
Another way to avoid such situations is to use smart constructors and not allow to create data types of `Monster` and `Knight` with constructors, only with your smart constructor function.